Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,442 @@
+    /// InvalidatePathCache button in the editor on the component.
+    ///
+    /// Collider's Transform changes are supported, but after changing the Rotation component the cache is going to be
+    /// invalid, and the user needs to invalidate it if they want to have a correct cache. If the collider is rotated,
+    /// non-uniform scale will distort the confiner.
+    /// 
+    /// The cache is NOT automatically invalidated (due to high computation cost every frame) if the contents
+    /// of the confining shape change (e.g. points get moved dynamically). In that case, we expose an API to
+    /// forcibly invalidate the cache so that it gets auto-recomputed next time it's needed.
+    /// </summary>
+    [SaveDuringPlay, ExecuteAlways]
+    public class CinemachineConfiner2D : CinemachineExtension
+    {
+        /// <summary>The 2D shape within which the camera is to be contained.</summary>
+        [Tooltip("The 2D shape within which the camera is to be contained.  " +
+                 "Can be a 2D polygon or 2D composite collider.")]
+        public Collider2D m_BoundingShape2D;
+
+        /// <summary>Damping applied automatically around corners to avoid jumps.</summary>
+        [Tooltip("Damping applied around corners to avoid jumps.  Higher numbers are more gradual.")]
+        [Range(0, 5)]
+        public float m_Damping = 0;
+
+        /// <summary>Draws Gizmos for easier fine-tuning.</summary>
+        [Tooltip("Draws Input Bounding Shape (black) and Confiner (cyan) for easier fine-tuning.")]
+        public bool m_DrawGizmos = true;
+        
+        /// <summary>
+        /// The confiner will correctly confine up to this maximum orthographic size. If set to 0, then this parameter is ignored and all camera sizes are supported. Use it to optimize computation and memory costs.
+        /// </summary>
+        [Tooltip("The confiner will correctly confine up to this maximum orthographic size. " +
+                 "If set to 0, then this parameter is ignored and all camera sizes are supported. " +
+                 "Use it to optimize computation and memory costs.")]
+        public float m_MaxOrthoSize;
+        
+        /// <summary>Invalidates cache and consequently trigger a rebake at next iteration.</summary>
+        public void InvalidatePathCache()
+        {
+            m_shapeCache.Invalidate();
+        }
+
+        /// <summary>Validates cache if it is invalid.</summary>
+        /// <param name="cameraAspectRatio">Aspect ratio of camera.</param>
+        /// <returns>Return true if cache is valid. False, otherwise.</returns>
+        public bool ValidatePathCache(float cameraAspectRatio)
+        {
+            return m_shapeCache.ValidateCache(
+                m_BoundingShape2D, m_MaxOrthoSize, m_confinerBaker, cameraAspectRatio, out _);
+        }
+        
+        private ConfinerOven m_confinerBaker = new ConfinerOven();
+        private VcamExtraState m_extra;
```

Comment:
Don't store this, it makes no sense.  Use the GetExtraState API from the base class

Reply:
I need to store it because the gizmos need it. GetExtraState<>() is protected, it needs to be protected internal for me to have access to it from the editor script. I did not want to change it. But we could do that too.

Is there a smarter way? :)